### ISSUE TITLE
System.ObjectDisposedException: The semaphore has been disposed.

### ISSUE THREAD

**Issue by datoml** [getsentry/sentry-dotnet#286]
Hello,

I am using the latest pre release version 2.0.0-beta 4 with aspnet core 3.
During the execution of my production code I get a lot off errors from the Sentry client itself.

This is the output of the log.

```
fail: Sentry.ISentryClient[0]
      An error occured when capturing the event System.ObjectDisposedException: The semaphore has been disposed.
         at System.Threading.SemaphoreSlim.CheckDispose()
         at System.Threading.SemaphoreSlim.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.SemaphoreSlim.Wait(TimeSpan timeout)
         at Sentry.Internal.BackgroundWorker.EnqueueEvent(SentryEvent event)
         at Sentry.SentryClient.DoSendEvent(SentryEvent event, Scope scope)
         at Sentry.SentryClient.CaptureEvent(SentryEvent event, Scope scope).
```

Has anyone an idea what causes this?
Thanks!

**Comment by bruno-garcia** [comment]
@datoml thanks for reporting it. And thanks a lot for running the preview bits.

This is odd. The error comes from the `SemaphoreSlim` being disposed. But `EnqueueEvent` has a guard to throw in case the `BackgroundWorker` is disposed:

https://github.com/getsentry/sentry-dotnet/blob/f6007f621d4935ca2570c0163ede944d7c86483f/src/Sentry/Internal/BackgroundWorker.cs#L60

A few things to confirm:  

1. you are registering `ISentryClient` or `IHub` yourself in the container?
2. Or newing up either `Hub` or `SentryClient`?
3. I see from the format but to confirm: It's just logging the internal errors due to `Debug=true`, right? This is not affecting your app in any way besides not reporting errors?
4. Do you see any other errors coming from `Sentry.` on your logs before this one comes in?

Last question is because the Semaphores get disposed here:

https://github.com/getsentry/sentry-dotnet/blob/f6007f621d4935ca2570c0163ede944d7c86483f/src/Sentry/Internal/BackgroundWorker.cs#L159

Which is when the worker look ends (a graceful shutdown/worker Dispose which doesn't seem to be the case as pointed out before) or an error in a block other than the call to the `Transport`:

https://github.com/getsentry/sentry-dotnet/blob/f6007f621d4935ca2570c0163ede944d7c86483f/src/Sentry/Internal/BackgroundWorker.cs#L126-L128 

That's because any exception there simply gets logged and suppressed:

https://github.com/getsentry/sentry-dotnet/blob/f6007f621d4935ca2570c0163ede944d7c86483f/src/Sentry/Internal/BackgroundWorker.cs#L136-L139

So the way I can see, that could've happened only if code that shouldn't throw, did, like here:

https://github.com/getsentry/sentry-dotnet/blob/f6007f621d4935ca2570c0163ede944d7c86483f/src/Sentry/Internal/BackgroundWorker.cs#L102
https://github.com/getsentry/sentry-dotnet/blob/f6007f621d4935ca2570c0163ede944d7c86483f/src/Sentry/Internal/BackgroundWorker.cs#L122

Or, the logger:

https://github.com/getsentry/sentry-dotnet/blob/f6007f621d4935ca2570c0163ede944d7c86483f/src/Sentry/Internal/BackgroundWorker.cs#L133

Additional questions:
1. Have you registered a custom logger? 
2. Or is this just `Sentry.AspNetCore`?
3. Do you use the Serilog or NLog integration?

**Comment by datoml** [comment]
Hello,
I hope I am able to answer your questions.

I registered the client like shown in the [ASP.NET Core integration](https://github.com/getsentry/sentry-dotnet#aspnet-core-integration) section.
```cs
public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
	WebHost.CreateDefaultBuilder(args)
		.UseStartup<Startup>()
		.UseKestrel()
		.UseSentry(args[0]);
```
The `args[0]` argument is the Sentry DNS I pass on start.

To send captured errors to Sentry I use the following syntax.
```cs
SentrySdk.CaptureException(ex);
SentrySdk.CaptureMessage(JsonConvert.SerializeObject(ex.Model), Sentry.Protocol.SentryLevel.Warning);
```

As you mentioned in point 3, it is not affecting my application. It only sends a lot of error messages to the console output.

I am also not able to see any error on the Sentry backend before these errors occur.

The funny thing is that this is only happening with the production code running inside my Docker container.
With the development build, there is no such error in the output.

Regarding your additional questions.
I am only using the build-in error logging provided by ASP.NET core with the `ILogger` service.
No custom logic or soever.

**Comment by artyomturkin** [comment]
Same problem 
- "Sentry.AspNetCore" Version="1.2.0"
- Runtime: Asp.Net Core 2.2

```
fail: Sentry.ISentryClient[0]
      An error occured when capturing the event System.ObjectDisposedException: The semaphore has been disposed.
         at System.Threading.SemaphoreSlim.CheckDispose()
         at System.Threading.SemaphoreSlim.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.SemaphoreSlim.Wait(TimeSpan timeout)
         at Sentry.Internal.BackgroundWorker.EnqueueEvent(SentryEvent event)
         at Sentry.SentryClient.DoSendEvent(SentryEvent event, Scope scope)
         at Sentry.SentryClient.CaptureEvent(SentryEvent event, Scope scope).
```

Sentry is injected like this:

In Program.cs
```csharp
public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
                .UseSentry()
                .UseStartup<Startup>();
```
And actual injection:

```csharp
public class DefaultRequestHandler : IGCVPRequestsHandler
{
    private readonly DataflowBlocks _dataflowBlocks;
    private readonly IHub _sentry;

    public DefaultRequestHandler(DataflowBlocks dataflowBlocks, IHub sentry)
    {
        _dataflowBlocks = dataflowBlocks;
        _sentry = sentry;
    }
}
```

And called like:
```csharp
_sentry.AddBreadcrumb("Hash checked", category: "checkwithHash", level: BreadcrumbLevel.Debug);

.........

_sentry.CaptureException(e);
```

**Comment by bruno-garcia** [comment]
I've tried to reproduce this locally but wasn't able.
Would you be able to share a repro?

**Comment by artyomturkin** [comment]
@bruno-garcia unfortunately I won't be able to share the repo, because primary functions rely on proprietary libraries we can not share.

Some more information:
- the problem presents itself after several hours of runtime
- average load is ~ 1 r/s (that should be reported to sentry)
- running in docker container

**Comment by bruno-garcia** [comment]
Thanks for sharing the details. I'll try to setup a lab to reproduce this.

I do use the SDK on [nugettrends.com](nugettrends.com) where it ran for months before a new deploy and it also runs on Docker. I ship the logs to `logz` and haven't seen any errors there yet.

**Comment by AlihanE** [comment]
@bruno-garcia I'm experiencing same trouble here.
I believe issue is appearing here.

> Last question is because the Semaphores get disposed here:
> sentry-dotnet/src/Sentry/Internal/BackgroundWorker.cs
> Line 159 in f6007f6
>  inSemaphore.Dispose(); 

We have unstable connection to sentry service sometimes and this problem appeared when sentry client started capturing several exceptions in one second in parallel threads. At the third time we had problem with our web service we were expecting this issue to shoot out and unfortunately our expectations were confirmed.

**Comment by bruno-garcia** [comment]
@AlihanE Thanks for letting us know. It's obvious a problem and these racy issues are sometimes hard to reproduce. I'll try to reproduce this again soon.
Sorry for the inconvenience.

**Comment by bruno-garcia** [comment]
I've pushed a fix and will release it as 2.0.0-beta5 once `zeus.ci` gets back online. Tomorrow it should be up.
Thank y'all for raising this.
